File: LargestPalindromeProduct3.cpp
#include<iostream>
#include<math.h> 

using namespace std;

class Solution {
public:
    int largestPalindrome(int n) {
    	if(n == 1) return 9;
    	int upper = pow(10, n) - 1, lower = upper / 10;
    	
    	long long possibleNum = (long long)upper * upper;
    	
    	int divider = pow(10, n);
    	int left = possibleNum / divider;
    	int right = possibleNum % divider;
    	
    	if(left == reverseNum(right)) return possibleNum % 1337;
    	
    	if(reverseNum(left) > right) left--;
    	
    	possibleNum = (long long)left * divider + reverseNum(left);
    	while(left > lower) {
    		for(long long i = upper; i * i >= possibleNum; i--) {
    			if(possibleNum % i == 0) return possibleNum % 1337;
			}
			left--;
			possibleNum = (long long)left * divider + reverseNum(left);
		}
    	
		return 0;
    }
    
    int reverseNum(int num) {
    	int r = 0;
    	for(; num > 0; r = r * 10 + num % 10, num /= 10);
    	return r;
	}
};

int main() {
	int n;
	cin>>n;
	
	Solution *solution = new Solution();
	cout<<solution->largestPalindrome(n);
	
	return 0;
}

File: LargestPalindromeProduct3.py
class Solution:
    def largestPalindrome(self, n: int) -> int:
        if n == 1:
            return 9
        for z in range(2, 2 * 9 * (10 ** n) - 1):
            left = 10 ** n - z
            right = int(str(left)[::-1])
            if z ** 2 - 4 * right < 0:
                continue
            x = 1 / 2 * (z + (z ** 2 - 4 * right) ** 0.5)
            y = 1 / 2 * (z - (z ** 2 - 4 * right) ** 0.5)
            if x.is_integer() or y.is_integer():
                return (10 ** n * left + right) % 1337
            # if (z ** 2 - 4 * right) ** 0.5 == int((z ** 2 - 4 * right) ** 0.5):
            #     return (10 ** n * left + right) % 1337


if __name__ == '__main__':
    n = int(input())
    solution = Solution()
    print(solution.largestPalindrome(n))
